File: workfloworchestrator/src/main/java/com/greenops/workfloworchestrator/ingest/handling/testautomation/CommandBuilder.java
package com.greenops.workfloworchestrator.ingest.handling.testautomation;

import org.apache.logging.log4j.util.Strings;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@Component
public class CommandBuilder {

    private List<String> commands;

    public CommandBuilder() {
        commands = new ArrayList<>();
    }

    public CommandBuilder createFile(String filename, String fileContents) {
        commands.add(Strings.join(List.of("echo", "'" + fileContents + "'", ">", filename), ' '));
        return this;
    }

    public CommandBuilder compile(String filename) {
        commands.add("chmod +x " + filename);
        return this;
    }

    public CommandBuilder ls() {
        commands.add("ls");
        return this;
    }

    public CommandBuilder cat(String filename) {
        commands.add("cat " + filename);
        return this;
    }

    public CommandBuilder executeFileContents(String fileContents) {
        commands.add(fileContents);
        return this;
    }

    public CommandBuilder executeExistingFile(String filename) {
        commands.add("./" + filename);
        return this;
    }

    public List<String> build() {
        return List.of(String.join("\n", commands));
    }
}

File: workfloworchestratorgo/ingest/handling/testautomation/commandbuilder.go
package testautomation

import "strings"

type CommandBuilder interface {
	CreateFile(fileName, fileContents string) CommandBuilder
	Compile(fileName string) CommandBuilder
	LS() CommandBuilder
	CAT(fileName string) CommandBuilder
	ExecuteFileContents(fileContents string) CommandBuilder
	ExecuteExistingFile(fileName string) CommandBuilder
	Build() []string
}

type commandBuilder struct {
	commands []string
}

func NewCommandBuilder() CommandBuilder {
	return &commandBuilder{commands: []string{}}
}

func (c *commandBuilder) CreateFile(fileName, fileContents string) CommandBuilder {
	cmd := "echo " + "'" + fileContents + "'" + " > " + fileName
	c.commands = append(c.commands, cmd)
	return c
}

func (c *commandBuilder) Compile(fileName string) CommandBuilder {
	cmd := "chmod +x " + fileName
	c.commands = append(c.commands, cmd)
	return c
}

func (c *commandBuilder) LS() CommandBuilder {
	c.commands = append(c.commands, "ls")
	return c
}

func (c *commandBuilder) CAT(fileName string) CommandBuilder {
	cmd := "cat " + fileName
	c.commands = append(c.commands, cmd)
	return c
}

func (c *commandBuilder) ExecuteFileContents(fileContents string) CommandBuilder {
	c.commands = append(c.commands, fileContents)
	return c
}

func (c *commandBuilder) ExecuteExistingFile(fileName string) CommandBuilder {
	c.commands = append(c.commands, "./"+fileName)
	return c
}

func (c *commandBuilder) Build() []string {
	return []string{strings.Join(c.commands, "\n")}
}
